storage/range.go

```diff
@@ -481,16 +486,30 @@ func (r *Range) AddCmd(args proto.Request, reply proto.Response, wait bool) erro
 // there are any overlapping commands already in the queue. Returns
 // the command queue insertion key, to be supplied to subsequent
 // invocation of cmdQ.Remove().
-func (r *Range) beginCmd(start, end proto.Key, readOnly bool) interface{} {
+func (r *Range) beginCmd(header *proto.RequestHeader, readOnly bool) interface{} {
 	r.Lock()
 	var wg sync.WaitGroup
-	r.cmdQ.GetWait(start, end, readOnly, &wg)
-	cmdKey := r.cmdQ.Add(start, end, readOnly)
+	r.cmdQ.GetWait(header.Key, header.EndKey, readOnly, &wg)
+	cmdKey := r.cmdQ.Add(header.Key, header.EndKey, readOnly)
 	r.Unlock()
 	wg.Wait()
+	// Update the incoming timestamp if unset. Wait until after any
+	// preceding command(s) for key range are complete so that the node
+	// clock has been updated to the high water mark of any commands
+	// which might overlap this one in effect.
+	if header.Timestamp.Equal(proto.ZeroTimestamp) {
+		header.Timestamp = r.rm.Clock().Now()
+	}
 	return cmdKey
 }
 
+// endCmd removes a pending command from the command queue.
+func (r *Range) endCmd(cmdKey interface{}) {
```

Comment:
We use `endCmd` on errors, but on success we access cmdQ.Remove directly (and do other stuff under the scope of the same lock). We should either make `endCmd` the only thing that calls `cmdQ.Remove` (just like `beginCmd` is the only thing that calls `cmdQ.Add`) or get rid of it and just inline the Lock/Remove operations.